### PR TITLE
Rework planning to be based off the resolve tree and not directly pkgs

### LINKED ISSUE #144: Can I have two different versions of the same Rust library in one Bazel project?
Hi! Can I have two different versions of the same Rust library in one Bazel project?
For example, some projects within one Bazel repository can depend on `futures-0.1.x`, but some can require `futures-0.3.x`. How to handle such use case using cargo-raze?

### LINKED ISSUE #241: When no target set, flate2 fails to compile with latest HEAD
Carrying on from https://github.com/google/cargo-raze/pull/212#issuecomment-705507497

To reproduce, checkout https://github.com/ankitects/cargo-raze-bug-examples and run repro.sh. Tested on a Mac.

### LINKED ISSUE #269: Erroneous aliases for async-global-executor-1.4.2
Repro:

Add a dependency on the crate async-global-executor-1.4.2 (https://github.com/async-rs/async-global-executor/blob/master/Cargo.toml).

That has this section in the `Cargo.toml` file:

```
[features]
default = ["async-io"]
tokio02 = ["tokio02-crate"]
tokio03 = ["tokio03-crate"]

# ...

[dependencies.tokio02-crate]
package = "tokio"
version = "^0.2"
optional = true
default-features = false
features = ["rt-core"]

[dependencies.tokio03-crate]
package = "tokio"
version = "^0.3"
optional = true
default-features = false
features = ["rt"]
```

However, the file generated by `cargo raze` (v0.6.1, installed a few moments ago) looks like:

```
# boilerplate omitted

# buildifier: leave-alone
rust_library(
    name = "async_global_executor",
    crate_type = "lib",
    deps = [
        "@raze__async_executor__1_3_0//:async_executor",
        "@raze__async_io__1_1_10//:async_io",
        "@raze__futures_lite__1_11_2//:futures_lite",
        "@raze__num_cpus__1_13_0//:num_cpus",
        "@raze__once_cell__1_4_1//:once_cell",
        "@raze__tokio__0_2_22//:tokio",
    ],
    srcs = glob(["**/*.rs"]),
    crate_root = "src/lib.rs",
    edition = "2018",
    rustc_flags = [
        "--cap-lints=allow",
    ],
    version = "1.4.2",
    tags = [
        "cargo-raze",
        "manual",
    ],
    crate_features = [
        "async-io",
        "default",
        "tokio02",
        "tokio02-crate",
    ],
    aliases = {
        "@raze__tokio__0_2_22//:tokio": "tokio03_crate",
    },
)
```

Note the `aliases` section at the bottom, and how `tokio_0_2_22` points to `tokio03_crate`, resulting in the following build error:

```
INFO: From Compiling Rust lib async_global_executor v1.4.2 (1 files):
error[E0433]: failed to resolve: use of undeclared type or module `tokio02_crate`
  --> external/raze__async_global_executor__1_4_2/src/lib.rs:69:22
   |
69 | static TOKIO02: Lazy<tokio02_crate::runtime::Handle> = Lazy::new(|| {
   |                      ^^^^^^^^^^^^^ use of undeclared type or module `tokio02_crate`

error[E0433]: failed to resolve: use of undeclared type or module `tokio02_crate`
  --> external/raze__async_global_executor__1_4_2/src/lib.rs:70:42
   |
70 |     let mut rt = tokio02_crate::runtime::Builder::new()
   |                                          ^^^^^^^ not found in `tokio02_crate::runtime`
```

Changing that `tokio03_crate` to `tokio02_crate` is enough for the build to succeed.

### LINKED ISSUE #270: Incorrect `rand` alias generation for `libsecp256k1` crate
Add `libsecp256k1 = "0.3.5"` to the dependencies section of the `Cargo.toml` and run `cargo raze`.

With v0.6.1 this generates `libsecp256k1-0.3.5.BUILD.bazel` with the following entry (hopefully irrelevant parts removed):

```
...

rust_library(
    name = "secp256k1",
    crate_type = "lib",

    ...
    
    aliases = {
        "@raze__rand__0_7_3//:rand": "rand_test",
    },
)
```

Note the `aliases` entry. This causes `rand` to appear as `rand_test` when building, and you get the following error:

```
% bazel build //cargo:libsecp256k1
...
INFO: From Compiling Rust lib secp256k1 v0.3.5 (15 files):
error[E0432]: unresolved import `rand`
  --> external/raze__libsecp256k1__0_3_5/src/lib.rs:34:5
   |
34 | use rand::Rng;
   |     ^^^^ use of undeclared type or module `rand`

error: aborting due to previous error
```

Edit the generated build file to remove or comment out the incorrect alias and the crate can build.

### PR DESCRIPTION
**Pull request by GregBowyer**
There are a range of subtle and irritating bugs that stem from
attempting to work with renames from the package set provided by cargo
metadata.

Fortunately, recent versions of cargo metadata provide a limited version
of the resolve nodes, both as the original node ids and as a newer set
providing both rename and targetting information.

As such we can (hopefully) simplify crate -> dep node resolution, remove
a bunch of subtle aliasing bugs.

Fixes #241, fixes #269, fixes #270, resolves #144, resolves #187

### CODE REVIEW

**Comment by GregBowyer** [comment]
@UebelAndre sorry if this stomps on anything you have in progress

**Comment by UebelAndre** [comment]
It undoubtably will 😢. While I do like the changes here, I would greatly appreciate it if I could get #276 through first. I've had to rebase and resolve conflicts a few times already ☹️ 

If I could just get it reviewed, I can promise to have it updated as fast as I possibly can to make sure it gets closed as soon as possible 😃

File: impl/src/planning/subplanners.rs
Comment (GregBowyer) [inline impl/src/planning/subplanners.rs:270]: This bit, and its devil children I dont fully like, it would be nice maybe to just grab the root contect without having to discover it through resolving deps.

I could not think of or find a way to do it, its probably ok for now, and maybe a better future approach might be to move to guppy as a dep

File: impl/src/planning/subplanners.rs
Comment (GregBowyer) [inline impl/src/planning/subplanners.rs:366]: I noticed in working with this, the targetting functions are a bit scattered and not really that reused.

As such I removed a bunch of the additional functions, smushing things down towards a single function, and in avoiding allocation-o-rama (which we do a bit frequently in cargo-raze) made it return an iterator.

The part for the generation of bazel conditions I moved into the template itself, as it seemed like a lot of code was written just to essentially prepend a string onto the platform targets.

File: impl/src/planning/subplanners.rs
Comment (GregBowyer) [inline impl/src/planning/subplanners.rs:427]: This wonderful part of the metadata gives us *both* _renames_ and _targets_. Irritatingly its still a little tricky to detect renames fully (we need this to finally deduce the aliases - see `is_renamed`)

If https://github.com/rust-lang/cargo/issues/7289 gets solved then a lot of the left-over rename detection code can get murdered

File: impl/src/planning/subplanners.rs
Comment (GregBowyer) [inline impl/src/planning/subplanners.rs:440]: The problem manifests as this

```toml
[dependencies]
bytes_new = { version = "0.3.0", package = "bytes" }
bytes_old = { version = "0.3.0", package = "bytes" }
```

Strictly this is an error. Right now cargo metadata will basically lose both `bytes` deps from the resolve graph :gruntle: but a `cargo build` will scream about having the same basic package with multiple names.

I dont have a good solution right now (I dont think we should go back to trying to essentially emulate the resolve graph like before), so I guess a todo :/

File: impl/src/planning/subplanners.rs
Comment (GregBowyer) [inline impl/src/planning/subplanners.rs:573]: NIT: Maybe add a comment in keeping with the other functions?

**Comment by GregBowyer** [comment]
> It undoubtably will cry. While I do like the changes here, I would greatly appreciate it if I could get #276 through first. I've had to rebase and resolve conflicts a few times already frowning_face
> 
> If I could just get it reviewed, I can promise to have it updated as fast as I possibly can to make sure it gets closed as soon as possible smiley

I can hold on your changes and work with that, I think we get #127 in then #276 and then see how bad that breaks this.

**Comment by UebelAndre** [comment]
Yeah, I can rebase on top of that when that goes in.

**Comment by UebelAndre** [comment]
https://github.com/google/cargo-raze/pull/276 is rebased on top of https://github.com/google/cargo-raze/pull/127 now

**Comment by GregBowyer** [comment]
@acmcarther thoughts?

**Comment by acmcarther** [comment]
Was holding off on review because the other PR is in review, but that's taking a while (because I'm picky or something).

I'll review this now.

**Comment by GregBowyer** [comment]
Nah we are all picky ;P

I am sure this one is going to need more work anyhow there are things I cant think of a better way for that I dont like

File: impl/src/planning/subplanners.rs
Comment (acmcarther) [inline impl/src/planning/subplanners.rs:564]: Does it make sense to precompute some of these new properties? I'm lost in the details here, but part of the idea with the crate catalog was to make this type of lookup more efficient.

**Review by acmcarther — CHANGES_REQUESTED** [review]
Not ideal that the formatter changes are co-mingled with the behavior changes. A bit harder to review.

I am so unqualified to review these PRs at this point. I don't remember much about the implementation and anyway it has changed a lot since I wrote significant code here.

Oh well

File: impl/src/planning.rs
Comment (GregBowyer) [reply impl/src/planning.rs:39]: Rustdoc thing, but I was noticing in the ther review most comments in carg-raze are f the form `/** comment */` and so I should make the comments that I think (for the consistency)

File: impl/src/planning/subplanners.rs
Comment (GregBowyer) [reply impl/src/planning/subplanners.rs:366]: Yeah, I kinda ignore it in cargo-raze because the tool runs for me a few times a week, but when calling three consecutive functions seems to produce vector and string of each other it gets a bit much ;)

**Comment by GregBowyer** [comment]
> Not ideal that the formatter changes are co-mingled with the behavior changes. A bit harder to review.
> 
> I am so unqualified to review these PRs at this point. I don't remember much about the implementation and anyway it has changed a lot since I wrote significant code here.
> 
> Oh well

I would only bother with the first commit in the PR, I can just pull the second commit (the reformatting) and we can do a round of that after all crazy PRs get merged

File: impl/src/planning/subplanners.rs
Comment (GregBowyer) [reply impl/src/planning/subplanners.rs:564]: Humm yeah might work, and might be a good way to make the find the root sillyness go away.

File: impl/src/planning/subplanners.rs
Comment (GregBowyer) [reply impl/src/planning/subplanners.rs:318]: From what I have seen it typically is providing its not too crazy but this is pretty much at the limit and might need to just be a function directly.

**Comment by UebelAndre** [comment]
Crazy PR has been merged. No more crazy PRs from me for the foreseeable future 😅

**Comment by UebelAndre** [comment]
In hindsight, this should have been merged before any crazy PRs 😞 Sorry about pushing for mine over this one.

@GregBowyer do you have bandwidth in the foreseeable future to revive this effort 😅?

**Comment by GregBowyer** [comment]
I have time to pick this up again

**Comment by GregBowyer** [comment]
Ok restarting the patch and the review.

This should be the changes (and just the changes) from the original code. The unit tests will be redone by me to cover the bugs.

There is one test that fails but that is just down to how the test itself is constructed than any logic failure.

I am going to try a few runs locally to see if it does the right thing on a live codebase.

@acmcarther @UebelAndre If we review and get this in can we either call this PR dead or not let it get overridden, I am get fatigued by constantly rewriting this logic (this is the 4th time of writing this now :S)

**Comment by UebelAndre** [comment]
> @acmcarther @UebelAndre If we review and get this in can we either call this PR dead or not let it get overridden, I am get fatigued by constantly rewriting this logic (this is the 4th time of writing this now :S)

@GregBowyer I don't think I understand your request but I'm happy to do whatever I can to make this easier for you 😄 I owe you for being so flexible with my large PR. Just let me know how I can help

File: impl/src/rendering/templates/partials/targeted_dependencies.template
Comment (UebelAndre) [inline impl/src/rendering/templates/partials/targeted_dependencies.template:5]: This should use a templated workspace name. ie:
https://github.com/google/cargo-raze/blob/eee9428916139e4329d25d5b1c42de120486463f/impl/src/rendering/templates/crate.BUILD.template#L10

File: impl/src/util.rs
Comment (UebelAndre) [inline impl/src/util.rs:441]: Why are these tests not necessary?

**Review by UebelAndre — CHANGES_REQUESTED** [review]

File: impl/src/rendering/templates/partials/targeted_dependencies.template
Comment (GregBowyer) [reply impl/src/rendering/templates/partials/targeted_dependencies.template:5]: yep noted when I was updating `Cargo.toml` to the latest

File: impl/src/util.rs
Comment (GregBowyer) [reply impl/src/util.rs:441]: The functions for doing targets were broken out into three steps that essentially needed each other to actually work.

I made it one function since the API was scattered, in the process the tests that worked on the intermetidate steps are harder (and probably dont actually need testing as a unit) to test.

As I see it the function should go to being `fn(cargo_target) -> bazel_target`, rather than juggling through several functions. I can beef up the tests for targeting if wanted to cover the testing gap.

**Comment by GregBowyer** [comment]
FWIW assuming I fixerate tests (and add the ones from the 3rd version of this PR) this works for us for our main `Cargo.toml`

**Comment by UebelAndre** [comment]
Hey @GregBowyer just FYI, I think https://github.com/google/cargo-raze/pull/363 introduced some conflicts but they're just Clippy fixes (which is now enabled in CI).

**Comment by GregBowyer** [comment]
@UebelAndre I have bumped to latest (untested), want to do me a review?

**Comment by UebelAndre** [comment]
Sure 😄 I'll take a look tonight

File: impl/src/planning.rs
Comment (UebelAndre) [inline impl/src/planning.rs:254]: nit: any reason not to go with `crate_ctx` (understanding `crate` is not usable)?

File: impl/src/planning/subplanners.rs
Comment (UebelAndre) [inline impl/src/planning/subplanners.rs:247]: I feel like the sanitize is worth keeping but wanted to share https://github.com/google/cargo-raze/commit/9d74090f28cf7d546414a0591205efa5219c9bf2 here in case that's relevant. I think `sanatize_indent` might need to be updated

File: impl/src/planning/subplanners.rs
Comment (UebelAndre) [inline impl/src/planning/subplanners.rs:294]:
```suggestion
    // Build a list of dependencies while addressing a potential allowlist of target triples
```

File: impl/src/planning/subplanners.rs
Comment (UebelAndre) [inline impl/src/planning/subplanners.rs:444]:
```suggestion
    // If https://github.com/rust-lang/cargo/issues/7289 gets solved then a lot of the left-over rename detection code can
```

This is optional but I personally prefer full links unless it's to the same repo as the project.

File: impl/src/planning/subplanners.rs
Comment (UebelAndre) [inline impl/src/planning/subplanners.rs:520]: Since writing this, was did you ever find out? I don't think it's a thing but that's likely a result of my limited experience 😅

File: impl/src/planning/subplanners.rs
Comment (UebelAndre) [inline impl/src/planning/subplanners.rs:596]: Do you think it'd be a lot of work to make this enum? It sounds like a far better solution and I don't think this is used too widely. Can punt but I think it'd be some great polish.

File: impl/src/util.rs
Comment (UebelAndre) [inline impl/src/util.rs:139]:
```suggestion
      allowlist
```

**Review by UebelAndre — CHANGES_REQUESTED** [review]
A couple of comments and suggestions but otherwise sounds pretty good. I'm going to need to take another day to page some of the nuances here back into my brain and do another pass but still pretty excited for the direction!

**Comment by UebelAndre** [comment]
Since migrating most things to Bazel, ensuring the examples are up to date has not been added to CI. Would you be able to run `bazel run //tools:examples_raze` and commit the results? I suspect a few things might change based these changes.

File: impl/src/context.rs
Comment (GregBowyer) [reply impl/src/context.rs:34]: Yeah I can kill it, I doubt it adds much but its probably not that hard to hand roll the `Hash` and `PartialEq`

File: impl/src/planning.rs
Comment (GregBowyer) [reply impl/src/planning.rs:254]: None, I was only mildly shortening the name that was already there.

File: impl/src/planning/subplanners.rs
Comment (GregBowyer) [reply impl/src/planning/subplanners.rs:247]: Oops I thought I factored that in when I rebased so good catch.

File: impl/src/planning/subplanners.rs
Comment (GregBowyer) [reply impl/src/planning/subplanners.rs:520]: I have never seen one in the wild although `-sys` crates are by convention.

I will pull it, I think if it happens it will be a curious bug in the future to figure out (even if such a thing existed would it be a host or target compile for instance)

**Comment by GregBowyer** [comment]
I will pull in the comments (all sane I think) make sure that test failure is not something sinister.

I think I should also take another scan through the bug list to see if there are any other bugs that this solves.

File: impl/src/util.rs
Comment (GregBowyer) [reply impl/src/util.rs:123]: Its to let rust know that `expression` is only needed used in the closure when we are filtering the targets, as such it needs to exist for the length of the iterator as bound by `'static` and `'a`. That is the `filter_map` takes ownership of `expression` and will not share it.

This is similar for `allowlist` which is given to the `filter` closure. Since we are not sharing these references we can move them where they are needed. It is important to note that this is not an optimisation but is actually the ownership rules.

File: impl/src/util.rs
Comment (UebelAndre) [inline impl/src/util.rs:54]: Haven't done a full review yet but are there any paths where this might require `Serialize` and `Deserialize` as well?

Comment (UebelAndre) [reply impl/src/util.rs:54]: Specifically asking in regards to https://github.com/google/cargo-raze/pull/415 but that's likely out of scope for that change and something I'll go update 😄

**Comment by GregBowyer** [comment]
No need it's not going towards the templates so won't serialise

On Fri, Apr 23, 2021, at 3:08 PM, UebelAndre wrote:
> 

> *@UebelAndre* commented on this pull request.

> 
> In impl/src/util.rs <https://github.com/google/cargo-raze/pull/282#discussion_r619518891>:

> > @@ -49,20 +51,33 @@ static SUPPORTED_PLATFORM_TRIPLES: &[&str] = &[
   "x86_64-unknown-freebsd",
 ];
 
+#[derive(Debug, PartialEq, Eq, Copy, Clone)]

File: impl/src/util.rs
Comment (UebelAndre) [reply impl/src/util.rs:54]: https://github.com/google/cargo-raze/pull/282#issuecomment-825969996

File: impl/src/planning/subplanners.rs
Comment (UebelAndre) [inline impl/src/planning/subplanners.rs:580]: nit: Can you add a small doc for this?

File: impl/src/planning/subplanners.rs
Comment (UebelAndre) [inline impl/src/planning/subplanners.rs:499]: nit: add a doc here as well

**Review by UebelAndre — CHANGES_REQUESTED** [review]
Just another small pass. I'll take some time over the weekend to do a thorough review. The thing I want to look into is the test coverage for this. What's your confidence level that there's enough here to prevent a regression? (again, need to look, there might be enough already but I don't have that context paged in my brain right now 😅)

File: impl/src/metadata.rs
Comment (UebelAndre) [inline impl/src/metadata.rs:441]: Was there an issue with this before?

File: impl/src/planning/subplanners.rs
Comment (UebelAndre) [inline impl/src/planning/subplanners.rs:131]: Is this not information that already lives (or could already live) in `CrateContext`? Admittedly that struct could use some cleaning but whether or not the crate is a workspace member seems like it makes sense to store in the context.
https://github.com/google/cargo-raze/blob/ee2effee5b6c95cfcfd5e496677ce61dcd35c472/impl/src/context.rs#L116-L153

File: impl/src/rendering/templates/partials/targeted_dependencies.template
Comment (UebelAndre) [reply impl/src/rendering/templates/partials/targeted_dependencies.template:5]:
```suggestion
          "@{{ rust_rules_workspace_name }}//rust/platform:{{platform_target}}",
```

File: impl/src/rendering/templates/partials/build_script.template
Comment (UebelAndre) [inline impl/src/rendering/templates/partials/build_script.template:77]:
```suggestion
          "@{{ rust_rules_workspace_name }}//rust/platform:{{platform_target}}",
```

File: impl/src/testing.rs
Comment (UebelAndre) [inline impl/src/testing.rs:228]: Another question about using a reference. Just curious.

**Review by UebelAndre — CHANGES_REQUESTED** [review]
Ok, a few more nits and one or two questions. But otherwise I think this is looking really good! 😄

**Comment by UebelAndre** [comment]
@acmcarther this is looking near completion. Don't know if you're around to review it but it'd be great to get in. If not, maybe @dfreese can take a look? It'd probably be good to get your eyes on it too just so you're aware of the changes (sorry for the ping otherwise 😅)

File: impl/src/context.rs
Comment (dfreese) [inline impl/src/context.rs:53]: I've been bitten by some nasty bugs where hash or comparison functions don't cover the all of the data in the struct.

Is there another way to structure this, perhaps by using map from target string to alias (or &DependencyAlias)?

File: impl/src/planning/subplanners.rs
Comment (dfreese) [reply impl/src/planning/subplanners.rs:461]: +1 to TODO(#cargo-raze-issue): referencing this error.

**Review by dfreese — COMMENTED** [review]
Haven't had a chance to make a full pass at this, but had one major question, based on past experience.

**Comment by UebelAndre** [comment]
@GregBowyer friendly ping. I feel this is so close to going in 😄

**Comment by UebelAndre** [comment]
@GregBowyer @dfreese another friendly ping 😅 Would love to get this merged and am wondering what outstanding action items are. Happy to help move this forward however I can.

**Comment by UebelAndre** [comment]
Friendly ping again. Happy to help however I can to get this in but would love to see this merged 😄

File: impl/src/planning/subplanners.rs
Comment (UebelAndre) [inline impl/src/planning/subplanners.rs:288]:
```suggestion
    let is_proc_macro = targets.iter().any(|target| target.kind == "proc-macro");
```

Aren't string literals fun? 😄

**Comment by UebelAndre** [comment]
@dfreese is this not something you could apply the suggestions for and accept as is? I think it's in a good spot.

**Comment by dfreese** [comment]
> @dfreese is this not something you could apply the suggestions for and accept as is? I think it's in a good spot.

https://github.com/google/cargo-raze/pull/282#discussion_r620890929

Not having all fields included as a part of a hash/comparison function causes some incredibly hard bugs to figure out, so I'd want to see if there's an alternative solution to that first.

**Comment by UebelAndre** [comment]
> > @dfreese is this not something you could apply the suggestions for and accept as is? I think it's in a good spot.
> 
> https://github.com/google/cargo-raze/pull/282#discussion_r620890929
> 
> Not having all fields included as a part of a hash/comparison function causes some incredibly hard bugs to figure out, so I'd want to see if there's an alternative solution to that first.

Is that not something you or I could patch in this PR or in a follow up? I think @GregBowyer has put fourth a huge amount of effort implementing this PR at least 3 times. I'd hate to have this reimplemented a 4th time and there are other changes I think should go in.

**Pull request closed without merging**
